File: algorithm/LeetCode/3SumClosest/3SumClosest.java
public class Solution {
    public int threeSumClosest(int[] nums, int target) {
         int len = nums.length;
         int ret = 0;
         int diff = Integer.MAX_VALUE;
         Arrays.sort(nums);
        for(int i = 0; i < len - 2; ++i){
            int beg = i + 1;
            int end = len - 1;
            while(beg < end){
                int tmp = nums[i] + nums[beg] + nums[end];
                int tmpDiff  = Math.abs(tmp - target);
                if(tmpDiff < diff){
                    ret = tmp;
                    diff = tmpDiff;
                }
                if(tmp < target)
                    beg++;
                else if(tmp > target)
                    end--;
                else
                    return target;
            }
        }   
        return ret;
    }
}

File: algorithm/LeetCode/3SumClosest/3SumClosest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int sz = nums.size();
        sort(nums.begin(), nums.end());//当然了，首先还是要排序的
        int diff = INT_MAX;//这个直接取INT_MAX,这样计算出来的diff一定是比这个要小的
        int tmpDiff;
        int ans;
        int beg, end;
        for (int i = 0; i < sz - 2; ++i){
            beg = i + 1, end = sz - 1;
            while (beg < end){
                int sum = nums[i] + nums[beg] + nums[end];
                if ((tmpDiff = abs(sum - target)) < diff){
                    diff = tmpDiff;
                    ans = sum;
                }
                if (sum > target){
                    end--;
                }
                else if (sum < target){
                    beg++;
                }
                else
                    return sum;//相等的话就直接返回了
            }
        }
        return ans;
    }
};